src/entities/group/group.controller.spec.ts

```diff
@@ -0,0 +1,267 @@
+    req = {
+      url: 'http://localhost:3000/api/groups',
+    } as Request;
+
+    res = ({
+      send: jest.fn().mockReturnThis(),
+      sendStatus: jest.fn().mockReturnThis(),
+      status: jest.fn().mockReturnThis(),
+    } as unknown) as Response;
+  });
+
+  afterEach(() => {
+    jest.clearAllMocks();
+  });
+
+  describe('getGroups', () => {
+    it('should call group service with custom limit if it is in query params', async () => {
+      req.url = 'http://localhost:3000/api/users?limit=2';
+
+      jest
+        .spyOn(groupService, 'getAll')
+        .mockImplementation(() => Promise.resolve([]));
+      await groupController.getGroups(req, res);
+
+      expect(groupService.getAll).toHaveBeenCalledWith(2);
+    });
+
+    it('should call res.send with list of groups', async () => {
+      const groups = (['group1', 'group2'] as unknown) as GroupModel[];
+
+      jest
+        .spyOn(groupService, 'getAll')
+        .mockImplementation(() => Promise.resolve(groups));
+      await groupController.getGroups(req, res);
+
+      expect(res.send).toHaveBeenCalledWith(groups);
+    });
+
+    it('should call res.send with empty array if no groups were found in DB', async () => {
+      const groups = ([] as unknown) as GroupModel[];
+
+      jest
+        .spyOn(groupService, 'getAll')
+        .mockImplementation(() => Promise.resolve(groups));
+      await groupController.getGroups(req, res);
+
+      expect(res.send).toHaveBeenCalledWith([]);
+    });
+
+    it('should call res.send with Internal Server Error if group service throws error', async () => {
+      const errorMsg = 'some error';
+
+      jest
+        .spyOn(groupService, 'getAll')
+        .mockImplementation(() => Promise.reject(new Error(errorMsg)));
+      await groupController.getGroups(req, res);
+
+      expect(res.sendStatus).toHaveBeenCalledTimes(1);
+    });
+  });
+
+  describe('getGroupById', () => {
+    it('should return 400 if no group id was supplied in params', async () => {
+      req.params = {
+        groupId: '',
+      };
+      await groupController.getGroupById(req, res);
+
+      expect(res.status).toHaveBeenCalledWith(ErrorCodes.BadRequest);
+      expect(res.send).toHaveBeenCalledTimes(1);
+    });
+
+    it('should return empty array if no group was found in db', async () => {
+      req.params = {
+        groupId: '1',
+      };
+
+      jest
+        .spyOn(groupService, 'getById')
+        .mockImplementation(() => Promise.resolve(null));
+
+      await groupController.getGroupById(req, res);
+
+      expect(res.send).toHaveBeenCalledWith([]);
+    });
+
+    it('should return group if it exists in db', async () => {
+      req.params = {
+        groupId: '1',
+      };
+
+      const group = new GroupModel();
+
+      jest
+        .spyOn(groupService, 'getById')
+        .mockImplementation(() => Promise.resolve(group));
+
+      await groupController.getGroupById(req, res);
+
+      expect(res.send).toHaveBeenCalledWith(group);
```

Comment:
Also it's good to check here and in other places that dependent service is called and called only once, something like this:
`expect(groupService.getById).toBeCalledTimes(1);`

Reply:
agree